enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,284 @@
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: example-service
+  namespace: some-namespace
+  annotations:
+    metallb.universe.tf/address-pool: example-pool
+    k8s.ovn.org/egress-service: '{"network-hint":172.19.0.0/24}'
+spec:
+  selector:
+    app: example
+  ports:
+    - name: http
+      protocol: TCP
+      port: 8080
+      targetPort: 8080
+  type: LoadBalancer
+```
+
+3. Advertise the service from the node in charge of the service's traffic. So far the service is "broken" - it is not reachable from outside the cluster and if the pods try to send traffic outside it would probably not come back as it is SNATed to an IP which is unknown.
+We create the advertisements targetting only the node that is in charge of the service's traffic using the `nodeSelectors` field, relying on ovn-k to label the node properly.
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: example-l2-adv
+  namespace: metallb-system
+spec:
+  ipAddressPools:
+  - example-pool
+  nodeSelectors:
+  - matchLabels:
+      k8s.ovn.org/example-service-some-namespace-host: ""
+---
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: example-l2-adv
+  namespace: metallb-system
+spec:
+  ipAddressPools:
+  - example-pool
+  nodeSelectors:
+  - matchLabels:
+      k8s.ovn.org/example-service-some-namespace-host: ""
+```
+While possible to create more advertisements resources for the `example-pool`, it is the user's responsibility to make sure that the pool is advertised only by advertisements targetting the node holding the `k8s.ovn.org/<svc-name>-<svc namespace>-host: ""` label - otherwise the traffic of the service will be broken.
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, I should be able to have functioning LoadBalancer services whose backing pods exit the cluster with the service's IP from an interface that belongs to a network I specify.
+
+### API Extensions
+A new annotation `k8s.ovn.org/egress-service` is supported for LoadBalancer services.
+
+### Test Plan
+
+- Unit tests coverage.
+- E2E coverage by creating a LoadBalancer service with the proper annotations and validating that:
+  - pod to pod traffic works.
+  - external client to service works.
+  - pod to external client works and is SNATed properly.
+
+
+### Risks and Mitigations
+
+- The solution might be a bit fragile as it relies on the user to configure the external advertisement of the service manually, with certain limitations.
+
+- Using a single node to handle all of the traffic of a given service might be a bottleneck, and we will also need to try electing nodes in a way that spreads handling these kind of services between them. Failover must also be taken care of in case a node handling a service falls.
```

Comment:
This is definitely something that will need to be addressed. It needs to be at least as resilient as egress IPs.